src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -332,6 +332,9 @@ private function createFirewall(ContainerBuilder $container, $id, $firewall, &$a
         // Provider id (take the first registered provider if none defined)
         if (isset($firewall['provider'])) {
             $defaultProvider = $this->getUserProviderId($firewall['provider']);
+            if (!in_array($defaultProvider, $providerIds, true)) {
+                throw new InvalidConfigurationException(sprintf('User provider "%s" used for firewall "%s" does not exist.', $firewall['provider'], $id));
```

Comment:
would that be more readable (similar below)?
`Invalid firewall "%s": user provider "%s" not found.`

Reply:
👍  should we use the very same message for the case where the provider is configured under a listener (handled below)? I think so